src/distributed_copy.c

```diff
@@ -0,0 +1,845 @@
+						errmsg("Copy in binary mode is not currently supported")));
+	}
+	PG_TRY();
+	{
+		char partitionType = PartitionType(tableId);
+		Oid intervalTypeId = (partitionType == HASH_PARTITION_TYPE) 
+			? INT4OID : partitionColumn->vartype;
+		bool useBinarySearch = (partitionType != HASH_PARTITION_TYPE);
+		int shardIndex = 0;
+
+		/* Lock all shards in shared mode */
+		shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+
+		shardCount = shardIntervalList->length;
+		shardIntervalCache = palloc0(shardCount * sizeof(ShardInterval*));
+		hashTokenIncrement = (uint32) (HASH_TOKEN_COUNT / shardCount);
+
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+			shardIntervalCache[shardIndex] = shardInterval;
+
+			if (partitionType == HASH_PARTITION_TYPE)
+			{
+				int32 shardMinHashToken = INT32_MIN + (shardIndex * hashTokenIncrement);
+				int32 shardMaxHashToken = shardMinHashToken + (hashTokenIncrement - 1);
+				if (shardIndex == (shardCount - 1))
+				{
+					shardMaxHashToken = INT32_MAX;
+				}
+				if (DatumGetInt32(shardInterval->minValue) != shardMinHashToken ||
+					DatumGetInt32(shardInterval->maxValue) != shardMaxHashToken)
+				{
+					useBinarySearch = true;
+				}
+			}
+			shardIndex += 1;
+		}
+		Assert(shardIndex == shardCount);
+
+		if (useBinarySearch)
+		{	
+			typeEntry = lookup_type_cache(intervalTypeId, TYPECACHE_CMP_PROC_FINFO);
+			compareFunction = &(typeEntry->cmp_proc_finfo);
+
+			qsort_arg(shardIntervalCache, shardCount, sizeof(ShardInterval*), 
+					  CompareShardIntervalsByMinHashToken, compareFunction);
+		}
+		while (true)
+		{
+			ShardInterval *shardInterval = NULL;
+			ShardId shardId = 0;
+			bool found = false;
+			int hashedValue = 0;
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno-1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+			partitionColumnValue = columnValues[partitionColumn->varattno-1];
+			if (partitionType == HASH_PARTITION_TYPE)
+			{
+				hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
+														  partitionColumnValue));
+				if (useBinarySearch)
+				{
+					partitionColumnValue = Int32GetDatum(hashedValue);
+				}
+			}
+			if (!useBinarySearch)
+			{
+				shardHashCode =
+					(int) ((uint32) (hashedValue - INT32_MIN) / hashTokenIncrement);
+				shardInterval = shardIntervalCache[shardHashCode];
+			}
+			else
+			{
```

Comment:
I agree, I actually only meant the binary search part, which should only have those 3 parameters.